openshift/machine-config-operator
# MCO-1144: Implement progressing/error boot images conditions on the MachineConfiguration object

**What I did**

- Moved feature gate detection outside the machine set boot image controller(MSBIC). The MSBIC will now only run when in tech preview, similar to the pinned image set controller.
- Removed the workqueue based system in the MSBIC, this helped with streamlining how the new conditions can be aggregated. This also removes the old metric based alert system.
- The MSBIC will now post two conditions: `MachineConfigurationBootImageUpdateProgressing` and `MachineConfigurationBootImageUpdateDegraded`, along with a message detailing how many machine resources are in that respective state.
- The operator degrades if the MSBIC posts an `MachineConfigurationBootImageUpdateDegraded=True`. 
- Added an e2e-techpreview test to verify that an misconfigured machineset causes a degrade on the operator

**How to test**

- Bring up a cluster in tech preview. Check the the conditions on the MachineConfiguration/cluster object, there will be no conditions as no machinesets have been opted-in
```
$ oc get machineconfiguration/cluster -o jsonpath-as-json='{.status.conditions}'
[]
```
- Apply a `MachineConfigurion` object that will opt-in some or all machinesets. In this case, I am applying one that opts in all machinesets: 
```
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  name: cluster
  namespace: openshift-machine-config-operator
spec:
  logLevel: Normal
  operatorLogLevel: Normal
  managedBootImages:
    machineManagers:
      - resource: machinesets
        apiGroup: machine.openshift.io
        selection:
          mode: All
```
The conditions should update to reflect this behavior:
```
$ oc get machineconfiguration/cluster -o jsonpath-as-json='{.status.conditions}'
[
    [
        {
            "lastTransitionTime": "2024-06-05T20:03:34Z",
            "message": "Reconciled 4 of 4 MAPI MachineSets | Reconciled 0 of 0 CAPI MachineSets | Reconciled 0 of 0 CAPI MachineDeployments",
            "reason": "MAPIMachinesetUpdated",
            "status": "False",
            "type": "BootImageUpdateProgressing"
        },
        {
            "lastTransitionTime": "2024-06-05T20:03:25Z",
            "message": "0 Degraded MAPI MachineSets | 0 Degraded CAPI MachineSets | 0 CAPI MachineDeployments",
            "reason": "MAPIMachinesetUpdated",
            "status": "False",
            "type": "BootImageUpdateDegraded"
        }
    ]
]
```
- Add an owner reference or something else to cause an error on a machineset. The owner reference can be anything, the content of it does not matter - as long as it exists. In this case, I added a dummy ownerReference such as this to one of the machinesets
```
$ oc get machinesets.machine.openshift.io/ci-ln-yqjiw12-72292-286vv-worker-a -n openshift-machine-api -o jsonpath-as-json='{.metadata.ownerReferences}'
[
    [
        {
            "apiVersion": "test",
            "kind": "test",
            "name": "test",
            "uid": "test"
        }
    ]
]
```
This will cause a degrade that should be visible in the conditions of the MachineConfiguration object and MCO's cluster operator object:
```
$ oc get machineconfiguration/cluster -o jsonpath-as-json='{.status.conditions}'
[
    [
        {
            "lastTransitionTime": "2024-06-05T20:08:14Z",
            "message": "Reconciled 3 of 4 MAPI MachineSets | Reconciled 0 of 0 CAPI MachineSets | Reconciled 0 of 0 CAPI MachineDeployments",
            "reason": "MAPIMachinesetUpdated",
            "status": "False",
            "type": "BootImageUpdateProgressing"
        },
        {
            "lastTransitionTime": "2024-06-05T20:08:15Z",
            "message": "1 Degraded MAPI MachineSets | 0 Degraded CAPI MachineSets | 0 CAPI MachineDeployments | Error(s): error syncing MAPI MachineSet ci-ln-yqjiw12-72292-286vv-worker-a: unexpected OwnerReference: test/test. Please remove this machineset from boot image management to avoid errors",
            "reason": "MAPIMachinesetUpdated",
            "status": "True",
            "type": "BootImageUpdateDegraded"
        }
    ]
]
$ oc get co/machine-config
NAME             VERSION                                                AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
machine-config   4.17.0-0.test-2024-06-05-170915-ci-ln-yqjiw12-latest   True        False         True       161m    Failed to resync 4.17.0-0.test-2024-06-05-170915-ci-ln-yqjiw12-latest because: bootimage update failed: 1 Degraded MAPI MachineSets | 0 Degraded CAPI MachineSets | 0 CAPI MachineDeployments | Error(s): error syncing MAPI MachineSet ci-ln-yqjiw12-72292-286vv-worker-a: unexpected OwnerReference: test/test. Please remove this machineset from boot image management to avoid errors
```
Removing the owner reference should remove the degrade on the CO object and update the MachineConfiguration conditions back to normal. Other errors should cause the same result, the ownerReferences path is just the fastest way to do it without affecting the cluster too much. For example, you could also:
1. Scale down the CVO and delete/mangle the coreos-bootimages configmap, so that the controller cannot find the update boot images anymore. In such a case, the controller will timeout after being unable to fetch and verify the configmap after 15 minutes. 
2. Mangle the machineset so it is an unexpected format, causing an error while unmarshalling.

- All e2e's should pass, including the new `TestBootImageDegradeCondition`.
